docs/documentation/upgrading/topics/keycloak/changes-24_0_0.adoc

```diff
@@ -163,3 +162,36 @@ To revert the change for an installation, use a custom Infinispan XML configurat
 ----
 <metrics names-as-tags="false" />
 ----
+
+= Changes to Password Hashing
+
+In this release we adapted the password hashing defaults to match the latest https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#pbkdf2[OWASP recommendations for Password Storage].
+
+The default password hashing algorithm has changed from `pbkdf2-sha256` to `pbkdf2-sha512`.
+
+Also, the number of default hash iterations for `pbkdf2` hashing algorithms changed:
+[%autowidth,cols="a,a,a"]
+|===
+| Algorithm | Old Iterations | New Iterations
+
+| pbkdf2-sha256 | 27.500 | 600.000
+| pbkdf2-sha512 | 30.000 | 210.000
+|===
+
+If a realm does not explicitly configure a password policy with `hashAlgorithm` and `hashIterations`, then
+the new configuration will take effect on the next password based login, or when a user password is created or updated.
+
+== Performance of new password hashing configuration
+
+Tests on a machine with an Intel i9-8950HK CPU (12) @ 4.800GHz yielded the following Ø time differences for hashing 1000 passwords:
+
+* with PBKDF2WithHmacSHA256 and 600.000 vs. 27.500 iterations: `+475ms` per password.
```

Comment:
One detail: Is it possible to add also what was the time for 27500 iterations VS new time with 600K iterations? For example, if before it took 1ms and now it takes 476 ms (475+1), it is much bigger "relative" difference than if it was 1000 ms before and now 1475 ms :-)

Reply:
@mposolda thanks for your suggestions. Here are the results from my comparisons:

```
Hashing 1000 password(s) with PBKDF2WithHmacSHA256
	Ø hashing duration with 27500 iterations: 20ms
	Ø hashing duration with 600000 iterations: 451ms
	Difference: +431 ms
Hashing 1000 password(s) with PBKDF2WithHmacSHA512
	Ø hashing duration with 30000 iterations: 33ms
	Ø hashing duration with 210000 iterations: 224ms
	Difference: +191 ms
```

I also included the `PBKDF2WithHmacSHA1`, but the speed difference is even more dramatic than the other algorithms:
```
Hashing 1 password(s) with PBKDF2WithHmacSHA1
	Ø hashing duration with 20000 iterations: 122ms
	Ø hashing duration with 1300000 iterations: 3114ms
	Difference: +2992 ms
```
Note that I had to reduce the number of repeated runs for PBKDF2WithHmacSHA1 from 1000 to 1, since the this algorithm runs incredibly slow.

Since `PBKDF2WithHmacSHA1` is not used by default and is imho rearely used in practice (speaking from over 200 customer projects) . I recommend that we deprecate `PBKDF2WithHmacSHA1` and just print a deprecation warning when the `Pbkdf2PasswordHashProviderFactory` is initialized to the log while either keeping the number of iterations at 20.000 or raise that number just slightly to a somewhat higher number which keeps the hashing < 1 second.  Otherwise users who use `Pbkdf2PasswordHashProviderFactory` might see a significant performance impact which could lead to CPU resource starvation.

Reply:
Thanks Thomas! Is it possible to add this info (the original time with previous number iterations VS the new time with new number of iterations) also to the `changes-24_0_0.adoc` file, so people are aware of the original time VS new time? 

As just the info about difference like `+475 ms` is not very useful IMO if you don't know what was the original time with the original number of iterations.

For `PBKDF2WithHmacSHA1` : I am fine with both possibilities (either keep 1300000 like you already did or keep original number and just print the WARNING in case this algorithm is used). Leaving to you to what you preferred.

Reply:
@mposolda thanks for your remarks!

I just updated the doc with your suggestions.

I also kept the updated iteration count for `Pbkdf2PasswordHashProviderFactory` (PBKDF2WithHmacSHA1) with provider id `pbkdf2` at 1.300.000 and choose to print a warning to the log the first time the provider is used.

The log message looks like:
```
2024-01-24 16:06:05,452 WARN  [org.keycloak.credential.hash.Pbkdf2PasswordHashProviderFactory] (executor-thread-9) Detected usage of password hashing provider 'pbkdf2'. The provider is no longer recommended, use 'pbkdf2-sha256' or 'pbkdf2-sha512' instead.
```
Note that we could also add the realm from session.getContext().getRealm().getName() to the log message, but I don't know if this information is always available.

AFAIK the password hashing provider `pbkdf2` is only used by legacy systems and in some test realms in the Keycloak project itself, however those test systems use a custom iteration count like 1 or 100k.
I added some guidance on how to reduce the number of hash iterations again for the legacy `pbkdf2` provider.